src/Akismet.php

```diff
@@ -0,0 +1,238 @@
+<?php
+
+/*
+ * This file is part of Flarum.
+ *
+ * For detailed copyright and license information, please view the
+ * LICENSE file that was distributed with this source code.
+ */
+
+namespace Flarum\Akismet;
+
+use GuzzleHttp\Client;
+use GuzzleHttp\Exception\GuzzleException;
+use Psr\Http\Message\ResponseInterface;
+
+class Akismet
+{
+    private $apiKey;
+    private $apiUrl;
+    private $flarumVersion;
+    private $extensionVersion;
+
+    private $params = [];
+
+    public function __construct(string $apiKey, string $homeUrl, string $flarumVersion, string $extensionVersion, bool $inDebugMode = false)
+    {
+        $this->apiKey = $apiKey;
+        $this->apiUrl = "https://$apiKey.rest.akismet.com/1.1";
+        $this->params['blog'] = $homeUrl;
+
+        $this->flarumVersion = $flarumVersion;
+        $this->extensionVersion = $extensionVersion;
+
+        if ($inDebugMode) {
+            $this->params['is_test'] = true;
+        }
+    }
+
+    public function isConfigured(): bool
+    {
+        return !empty($this->apiKey);
+    }
+
+    /**
+     * @param  string  $type  e.g. comment-check, submit-spam or submit-ham;
+     * @throws GuzzleException
+     */
+    protected function sendRequest(string $type): ResponseInterface
+    {
+        $client = new Client();
+        return $client->request('POST', "$this->apiUrl/$type", [
+            'headers'     => [
+                'User-Agent' => "Flarum/$this->flarumVersion | Akismet/$this->extensionVersion",
+            ],
+            'form_params' => $this->params,
+        ]);
+    }
+
+    /**
+     * @throws GuzzleException
+     */
+    public function checkSpam(): array
+    {
+        $response = $this->sendRequest('comment-check');
+
+        return [
+            'isSpam' => $response->getBody()->getContents() === 'true',
+            'proTip' => $response->getHeaderLine('X-akismet-pro-tip'),
+        ];
+    }
+
+    /**
+     * @throws GuzzleException
+     */
+    public function submitSpam()
+    {
+        $this->sendRequest('submit-spam');
+    }
+
+    /**
+     * @throws GuzzleException
+     */
+    public function submitHam()
+    {
+        $this->sendRequest('submit-ham');
+    }
+
+    private function getImmutableWithParam(string $key, $value): Akismet
+    {
+        $clone = $this;
+        $clone->params[$key] = $value;
+
+        return $clone;
+    }
```

Comment:
Maybe instead of this util, we should just call `withParams`?

Reply:
@askvortsov1 The `withParams` method takes an array as a parameter and is there so that you can pass your own parameters that are not included in the methods. E.g. if one day new parameters are added to the Akismet API and if the author of an extension that uses this class wants to use them then he will not have to wait for the extension to be updated.

Reply:
Hmmm, but actually if we make the `withParam` function public, then the `withParams` method will not be needed

Reply:
My point is that since `getImmutableWithParam` can be implemented via `withParams`:

```
    private function getImmutableWithParam(string $key, $value): Akismet
    {
        return $this->withParams([$key => $value]);
    }
```

it's redundant, and we could just use `withParams`. Or vice versa, and have the singular method be public.